### PR TITLE
dxf: handle modifying task concurrency in scheduler

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #57497

Problem Summary:

### What changed and how does it work?
handle modify task concurrency in scheduler

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/storage/task_state.go [hunk 1]
```diff
@@ -199,6 +199,41 @@ func (mgr *TaskManager) ModifyTaskByID(ctx context.Context, taskID int64, param
...
+		}
+		if se.GetSessionVars().StmtCtx.AffectedRows() == 0 {
+			// might be handled by other owner nodes, skip.
+			return nil
+		}
+		// subtask in final state are not changed.
+		// subtask might have different concurrency later, see TaskExecInfo, we
+		// need to handle it too, but ok for now.
+		_, err = sqlexec.ExecSQL(ctx, se.GetSQLExecutor(), `
+			update mysql.tidb_background_subtask
```
Comment (lance6716) [inline pkg/disttask/framework/storage/task_state.go:228]: Just learning, how do we let subtask executor know the concurrency is changed?

**Review by lance6716 — COMMENTED** [review]
will review unit test files later

File: pkg/disttask/framework/scheduler/interface.go [hunk 2]
```diff
@@ -49,10 +49,14 @@ type TaskManager interface {
 	RevertedTask(ctx context.Context, taskID int64) error
 	// PauseTask updated task state to pausing.
 	PauseTask(ctx context.Context, taskKey string) (bool, error)
-	// PausedTask updated task state to paused.
+	// PausedTask updated task state to 'paused'.
 	PausedTask(ctx context.Context, taskID int64) error
 	// ResumedTask updated task state from resuming to running.
 	ResumedTask(ctx context.Context, taskID int64) error
+	// ModifiedTask tries to update task concurrency and meta, and update state
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/interface.go:56]: 1. `state` is not a param, it's for changing state back to the state before modify, see this state machine https://github.com/pingcap/tidb/pull/57557/files#diff-9cbb4aa7310f95ed10a52e14e392efa29e0356b179e85efec38db8fde5dca9f3R57-R67. for priority, yes
2/3. see above link

File: pkg/disttask/framework/integrationtests/modify_test.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		subtaskCh <- struct{}{}
+		task2Base := testutil.WaitTaskDone(c.Ctx, t, theTask.Key)
+		require.Equal(t, proto.TaskStateSucceed, task2Base.State)
+		checkSubtaskConcurrency(t, c, theTask.ID, map[proto.Step]int{
+			proto.StepOne: 7,
+			proto.StepTwo: 7,
+		})
+	})
+
+	t.Run("modify pending task concurrency, but other owner already done it", func(t *testing.T) {
```
Comment (lance6716) [inline pkg/disttask/framework/integrationtests/modify_test.go:167]: The test purpose is "..but other owner already done it". Should we trigger ModifiedTask after WaitTaskDone and check the concurrency is not changed?

**Review by lance6716 — APPROVED** [review]

File: pkg/disttask/framework/integrationtests/modify_test.go (hunk 3 above)
Comment (D3Hunter) [reply pkg/disttask/framework/integrationtests/modify_test.go:167]: > Should we trigger ModifiedTask after WaitTaskDone

the scenario is 
- task change to `modifying` state
- owner A tried to do modify work, but before it start, another owner B finished the modify
- owner A does nothing

> check the concurrency is not changed?

on owner A, we check by `AffectedRows() == 0`

**Review by tangenta — APPROVED** [review]

**Pull request merged**